### PR TITLE
[Move] transaction builder codegen in python (and Rust)

### PR DESCRIPTION
**Pull request by matbd**
## Motivation

* Generate some python code equivalent to what is currently generated by Rust macros in the `transaction-builder` crate.

* For testing purposes, I also generate Rust transaction builders from scratch.

Note that the ABI described by macros in `language/transaction-builder` seems currently out of date compared to the Move code.

## Test Plan

```
# Spit out the python code
cargo run -p transaction-builder-generator

# Verify that the generated Rust code compiles
cargo test -p transaction-builder-generator

# Run test on the python codegen (not supported by CI)
cargo test -p transaction-builder-generator -- --ignored 
```

## Related PRs

#4367

### CODE REVIEW

File: language/transaction-builder-generator/src/python3.rs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+pub fn output(out: &mut dyn Write, abis: &[ScriptABI]) -> Result<()> {
+    output_preambule(out)?;
+    for abi in abis {
+        output_builder(out, abi)?;
+    }
+    Ok(())
+}
+
+fn output_preambule(out: &mut dyn Write) -> Result<()> {
```
Comment (sblackshear) [inline language/transaction-builder-generator/src/python3.rs:17]:
```suggestion
fn output_preamble(out: &mut dyn Write) -> Result<()> {
```

File: language/transaction-builder-generator/src/python3.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        U8 => "st.uint8".into(),
+        U64 => "st.uint64".into(),
+        U128 => "st.uint128".into(),
+        Address => "libra.AccountAddress".into(),
+        Vector(type_tag) => match type_tag.as_ref() {
+            U8 => "bytes".into(),
+            _ => format!("typing.Sequence[{}]", quote_type(type_tag.as_ref())),
+        },
+
+        Struct(_) | Signer => panic!("not supported"),
```
Comment (sblackshear) [inline language/transaction-builder-generator/src/python3.rs:120]: Might want this error message to imply that this is banned by Move--the current message sounds like this just might not be feature supported by the codegen.

Comment (sblackshear) [reply language/transaction-builder-generator/src/python3.rs:120]: Same with other "not supported"'s on `Struct/Signer`

File: language/transaction-builder-generator/src/rust.rs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+pub fn output(out: &mut dyn Write, abis: &[ScriptABI]) -> Result<()> {
+    output_preambule(out)?;
+    for abi in abis {
+        output_builder(out, abi)?;
+    }
+    Ok(())
+}
+
+fn output_preambule(out: &mut dyn Write) -> Result<()> {
```
Comment (sblackshear) [inline language/transaction-builder-generator/src/rust.rs:17]:
```suggestion
fn output_preamble(out: &mut dyn Write) -> Result<()> {
```

File: language/transaction-builder-generator/src/python3.rs (hunk 1 above)
Comment (matbd) [reply language/transaction-builder-generator/src/python3.rs:17]: haha that's the right spelling but not in English :)

File: language/transaction-builder-generator/src/python3.rs (hunk 2 above)
Comment (matbd) [reply language/transaction-builder-generator/src/python3.rs:120]: Sure, I can do that.

For the records, I'm still curious why we allow complex types as type arguments but not complex values as regular arguments.

**Comment by thefallentree** [comment]
cc @ssinghaldev

File: language/transaction-builder-generator/src/python3.rs (hunk 2 above)
Comment (sblackshear) [reply language/transaction-builder-generator/src/python3.rs]: The following invariant is important for Move's security model: A record type `R` declared in module `M` can only be created by a `Pack` instruction inside a procedure of `M`.  If record values can be passed into a transaction script, this invariant doesn't hold.

The use of types is not privileged in the same--any module can declare a procedure or type that uses any type declared in another module (as long it doesn't create a circular dependency).

Comment (sblackshear) [reply language/transaction-builder-generator/src/python3.rs]: Said differently: producing a value of a particular type may be privileged, but producing a type of a particular kind never is.

Comment (matbd) [reply language/transaction-builder-generator/src/python3.rs]: Thanks for the explanation! Based on what you're saying, nothing prevents us from having arbitrarily nested objects in TransactionArgument as long as they are benign enough (say vectors of vectors). So I understand we should probably plan for it in terms of code generation and LCS stack overflow.

**Comment by thefallentree** [comment]
looks great! I assume the generated file will be checked in and checked similar how yaml spec is checked ?

**Pull request closed without merging**